sky/clouds/utils/scp_utils.py

```diff
@@ -442,3 +442,162 @@ def start_instance(self, vm_id):
+
+            services = []
+            for port in ports:
+                services.append({'serviceType': 'TCP', 'serviceValue': port})
+
+            if direction == 'IN':
+                target_address = 'sourceIpAddresses'
+            else:
+                target_address = 'destinationIpAddresses'
+
+            request_body = {
+                'ruleDirection': direction,
+                'services': services,
+                target_address: ['0.0.0.0/0'],
+                'ruleDescription': 'skyserve rule'
+            }
+
+            return self._post(url, request_body)
+
+    def list_firewall_rules(self, firewall_id):
+        url = f'{API_ENDPOINT}/firewall/v2/firewalls/{firewall_id}/rules'
+        return self._get(url)
+
+    def _check_existing_firewall_rule(self, firewall_id, internal_ip, direction,
+                                      ports):
+        response = self.list_firewall_rules(firewall_id)
+        rules = []
+        for rule in response:
+            if direction == 'IN':
+                if internal_ip == rule['destinationIpAddresses'][0]:
+                    rules.append(rule)
+            else:
+                if internal_ip == rule['sourceIpAddresses'][0]:
+                    rules.append(rule)
+        for rule in rules:
+            port_list = ','.join(rule['tcpServices'])
+            port = ','.join(ports)
+            if port == port_list:
+                return False
+        return True
+
+    def add_new_firewall_rule(self, firewall_id, internal_ip, direction, ports):
+        if self._check_existing_firewall_rule(firewall_id, internal_ip,
+                                              direction, ports):
+            url = f'{API_ENDPOINT}/firewall/v2/firewalls/{firewall_id}/rules'
+
+            services = []
+            for port in ports:
+                services.append({'serviceType': 'TCP', 'serviceValue': port})
+
+            if direction == 'IN':
+                source_ip = '0.0.0.0/0'
+                destination_ip = internal_ip
+            else:
+                source_ip = internal_ip
+                destination_ip = '0.0.0.0/0'
+
+            request_body = {
+                'sourceIpAddresses': [source_ip],
+                'destinationIpAddresses': [destination_ip],
+                'services': services,
+                'ruleDirection': direction,
+                'ruleAction': 'ALLOW',
+                'isRuleEnabled': True,
+                'ruleLocationType': 'FIRST',
+                'ruleDescription': 'skyserve rule'
+            }
+            return self._post(url, request_body)
+
+    def wait_firewall_rule_complete(self, firewall_id, rule_id):
+        while True:
+            time.sleep(5)
+            rule_info = self.get_firewall_rule_info(firewall_id, rule_id)
+            if rule_info['ruleState'] == 'ACTIVE':
+                break
+        return
```

Comment:
```suggestion
                return
```
nit. also, should we have a max retry threshold and raise some error if encountered?

Reply:
"open_ports" in sky/provision/scp/instance.py that calls "wait_firewall_rule_complete" has try clause, it is okay.